Update parsing of alerts to be able to parse directionId and routeType

evansiroky opened the pull request:
This PR updates some stuff with the parsing of alerts to be able to handle the directionId (at the root level) and also the routeType. Some additional refactoring was done to improve the code. The biggest change is replacing the `src/main/java/com/google/transit/realtime/GtfsRealtime.java` file thus updating it to the latest version. This file is copied verbatim from  https://github.com/MobilityData/gtfs-realtime-bindings/blob/master/java/src/main/java/com/google/transit/realtime/GtfsRealtime.java. As a result of updating to the latest gtfs-realtime-bindings the alert type of MODIFIED was lost. I couldn't find any info on this alert type in the GTFS-RT spec, so I'm not sure what it is. I ended up removing all code paths related to it.

evansiroky committed:
Update parsing of alerts to be able to parse directionId and routeType

evansiroky committed:
Update to latest osmpbf package

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/updater/alerts/AlertsUpdateHandler.java:
```diff
@@ -149,12 +155,15 @@ private void handleAlert(String id, GtfsRealtime.Alert alert) {
+            + (informed.hasRouteType() ? informed.getRouteType() : " null ") + " "
+            + (informed.hasStopId() ? informed.getStopId() : " null ") + " "
             + (informed.hasTrip() && informed.getTrip().hasTripId() ?
                 informed.getTrip().getTripId() : " null ");
```
Can this long string concat become a `String.join`, with `" "` as the separator?

evansiroky replied:
Refactored in d074833db78929d4776ad0ac37724965657cead7

landonreed commented on src/main/java/org/opentripplanner/routing/alertpatch/AlertPatch.java:
```diff
@@ -61,6 +61,11 @@
+    /**
+     * Route type of the route(s) this alert concerns, set to -1 if no route type.
+     */
+    private int routeType = -1;
```
Would it be worth using a constant for this -1 value since it gets used elsewhere in this file? E.g. on line 256 and in AlertsUpdateHandler (line 123).

evansiroky replied:
Refactored in d074833db78929d4776ad0ac37724965657cead7

landonreed commented on src/main/java/org/opentripplanner/routing/alertpatch/AlertPatch.java:
```diff
@@ -131,79 +133,40 @@ public void apply(Graph graph) {
+    /**
+     * Returns true if the trip pattern would NOT be affected by this alert patch.
+     */
+    private boolean isNotApplicableTripPattern(TripPattern tripPattern, Agency alertAgency) {
```
Non-blocking Nit (NBN?): maybe rename method to `tripPatternIsUnaffectedByAlert`. Just a touch more readable.

evansiroky replied:
I actually originally named the variable similar to that, but decided to prefix with `is` because it seemed like it might follow some kind of naming convention. And in fact there is something I can reference to: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members?redirectedfrom=MSDN

binh-dam-ibigroup replied:
How about negating the function and rename to `appliesToTripPattern`?

landonreed commented on src/main/java/org/opentripplanner/routing/alertpatch/AlertPatch.java:
```diff
@@ -131,79 +133,40 @@ public void apply(Graph graph) {
+     * Returns true if the trip pattern would NOT be affected by this alert patch.
+     */
+    private boolean isNotApplicableTripPattern(TripPattern tripPattern, Agency alertAgency) {
+        // make sure direction matches. Not sure why there is a string direction value still.
```
I don't fully understand this comment. Are you asking about why the direction field is of the type `String` and not  `int`?

evansiroky replied:
yeah. This String direction field doesn't appear to be used in the OTP 2 branch anyways.

evansiroky replied:
See https://github.com/ibi-group/OpenTripPlanner/pull/55#discussion_r632825705

landonreed commented on src/main/java/org/opentripplanner/routing/alertpatch/AlertPatch.java:
```diff
@@ -61,6 +61,11 @@
      */
     private int directionId = -1;
```
This is a comment for lines 49-51. I noticed this weirdness in the javadoc:

https://github.com/ibi-group/OpenTripPlanner/blob/946da1d32cfbe87deb8476fc127266190a234d05/src/main/java/org/opentripplanner/routing/alertpatch/AlertPatch.java#L49-L52

evansiroky replied:
ok, that's a good bit more context. It turns out there aren't any codepaths to set this anyways, so I went ahead and removed it altogether in e2bb4b8b2ab2c9b846496dcc76ae0a039ca59102

landonreed commented on src/main/java/org/opentripplanner/routing/alertpatch/AlertPatch.java:
```diff
@@ -84,17 +89,17 @@ public void setId(String id) {
-    public void apply(Graph graph) {
-        Agency agency = null;
+    public void applyOrRemove(Graph graph, Consumer<Edge> applyOrRemoveFn) {
+        Agency alertAgency = null;
```
This variable name change feels a little weird only because the other entities (route and trip) weren't updated similarly. Is there a reason for that?

evansiroky replied:
I forget, so I changed it back in d074833db78929d4776ad0ac37724965657cead7

landonreed commented on src/main/java/org/opentripplanner/routing/alertpatch/AlertPatch.java:
```diff
@@ -268,6 +231,10 @@ public int getDirectionId() {
     }
 
+    public boolean hasDirectionId() {
+        return directionId != -1;
```
Same comment as https://github.com/ibi-group/OpenTripPlanner/pull/55/files#r632083309 (i.e., would it be worth using a constant for this -1 value?).

evansiroky replied:
Refactored in d074833db78929d4776ad0ac37724965657cead7

landonreed commented on src/main/java/org/opentripplanner/updater/stoptime/TimetableSnapshotSource.java:
```diff
@@ -230,9 +230,11 @@ public void applyTripUpdates(final Graph graph, final boolean fullDataset, final
+// this come from and is it still needed/or used anywhere?
+//                    case MODIFIED:
+//                        applied = validateAndHandleModifiedTrip(graph, tripUpdate, feedId, serviceDate);
+//                        break;
```
Yea, this is confusing. I would say maybe we just remove entirely? I'm not really sure though. I did a very brief search and found `MODIFIED` mentioned in https://github.com/opentripplanner/OpenTripPlanner/issues/2780 with a link to to this coderef:

https://github.com/opentripplanner/OpenTripPlanner/blob/de5454238b1132e55e5f22e11c287e58f506a57d/src/main/java/org/opentripplanner/updater/stoptime/TimetableSnapshotSource.java#L302-L305

Oh, haha, that's this file 😅 

Yea, this does feel like a stubbed out thing that never really got finished/accepted into the GTFS-rt spec? https://github.com/opentripplanner/OpenTripPlanner/blame/0ca80a2bd25d6c6a437838e01d1a5d31cfe29822/src/main/java/org/opentripplanner/updater/stoptime/TimetableSnapshotSource.java#L621-L624

evansiroky replied:
Well, this comment seems to explain things: https://github.com/ibi-group/OpenTripPlanner/blob/78639646c0ba61db359745215bce6df0ed30d399/src/main/java/com/google/transit/realtime/GtfsRealtime.java#L7-L25

The latest gtfs-rt-bindings include a deprecated field of REPLACEMENT with the same value, so I went ahead and changed things to use that in d801a8ffbd6db673458094a7ac9b9a6dbe92006e.

binh-dam-ibigroup reviewed (approved):
LGTM, just wondering whether a unit test might be useful for the new logic in `AlertsUpdateHandler`, even though `directionId` and `routeType` are ultimately sent to `patch`.

landonreed reviewed (changes requested):
A couple of minor things, but I do think this could really use some tests. Let me or @binh-dam-ibigroup if you need any extra hands for that!

evansiroky committed:
Remove unused direction field from AlertPath

evansiroky committed:
Uncomment and update some GTFS-RT changes

This now replaces the TripDescriptor.ScheduleRelationship.MODIFIED with TripDescriptor.ScheduleRelationship.REPLACEMENT to be compatible with the latest GTFS-RT bindings

evansiroky committed:
Address PR review comments

See https://github.com/ibi-group/OpenTripPlanner/pull/55

evansiroky committed:
Bring in and adapt some changes that sdjacobs made

See https://github.com/opentripplanner/OpenTripPlanner/pull/2607

evansiroky commented:
I have gone ahead and brought in some of the changes that @sdjacobs made in https://github.com/opentripplanner/OpenTripPlanner/pull/2607 that included some requested tests. Please review and approve if this is sufficient.

br648 reviewed (approved):
LGTM. All unit tests pass (on Windows).

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/routing/alertpatch/AlertPatchTest.java:
```diff
@@ -184,4 +184,97 @@ public void testRouteNotePatch() {
+
+    public void testRouteTypeAlert() {
+
+        // on route 18, direction 1 (N) is E -> D, 0 (S) is D -> E
```
Comment needs to match the test (route type instead of direction)... Also I am not sure to follow the logic, is a particular direction id implied?

evansiroky replied:
I added some more comments and moved this one around in 95788c8c4b0c18065885ca7bd9af7e6c729e1bbf.

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/updater/alerts/AlertsUpdateHandlerTest.java:
```diff
@@ -0,0 +1,91 @@
+
+    @Test
+    public void testAlertWithTimePeriod() {
+        GtfsRealtime.Alert alert = GtfsRealtime.Alert.newBuilder()
```
Can the content be refactored as follows?
```java
@Test
public void testAlertWithTimePeriod() {
    testAlertParameters(10, 20);
    // testAlertParameters(10, null), etc;
}
```

evansiroky replied:
Refactored in 95788c8c4b0c18065885ca7bd9af7e6c729e1bbf

binh-dam-ibigroup reviewed (approved):
I think this should work. Conditional approval on addressing the comment regarding testing route types.

evansiroky committed:
Set dateTime of request in tests to make sure tests run consistently

evansiroky committed:
Address PR review comments

See https://github.com/ibi-group/OpenTripPlanner/pull/55
